File: Ans1/data1.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import find_peaks


data_1 = np.loadtxt('Data_1.txt')


def find_peaks_and_minima(data):
  
    peaks, _ = find_peaks(data)
    
   
    minima, _ = find_peaks(-data)
    
    return peaks, minima


peaks_1, minima_1 = find_peaks_and_minima(data_1)
peaks_2, minima_2 = find_peaks_and_minima(data_2)

plt.figure(figsize=(12, 6))


plt.subplot(1, 2, 1)
plt.plot(data_1, label='Data_1')
plt.scatter(peaks_1, data_1[peaks_1], color='red', label='Maxima')
plt.scatter(minima_1, data_1[minima_1], color='blue', label='Minima')
plt.xlabel('Index')
plt.ylabel('Value')
plt.title('Data_1 Peaks and Minima')
plt.legend()



# Show the plots
plt.tight_layout()
plt.show()

File: Ans1/data1.c
#nclude <stdio.h>
#include <stdlib.h>

void find_peaks(double *data, int size) {
    printf("Index of Maxima:\n");
    for (int i = 1; i < size - 1; i++) {
        if (data[i] > data[i-1] && data[i] > data[i+1]) {
            printf("Maxima at index %d: %lf\n", i, data[i]);
        }
    }

    printf("Index of Minima:\n");
    for (int i = 1; i < size - 1; i++) {
        if (data[i] < data[i-1] && data[i] < data[i+1]) {
            printf("Minima at index %d: %lf\n", i, data[i]);
        }
    }
}

int main() {
    FILE *file = fopen("Data_1.txt", "r"); // Assuming 'Data_1.txt' is the filename
    if (file == NULL) {
        printf("Error opening file.\n");
        return 1;
    }

    int size = 0;
    fscanf(file, "%d", &size); // Assuming first line contains size

    double *data = (double*) malloc(size * sizeof(double));
    for (int i = 0; i < size; i++) {
        fscanf(file, "%lf", &data[i]);
    }
    fclose(file);

    find_peaks(data, size);
    free(data);

    return 0;
}
